src/components/shop-app.js

```diff
@@ -0,0 +1,489 @@
+          <iron-selector role="navigation" class="drawer-list" selected="${categoryName}" attr-for-selected="name">
+            ${repeat(categories, category => html`
+              <a name="${category.name}" href="/list/${category.name}">${category.title}</a>
+            `)}
+          </iron-selector>
+        </app-drawer>
+      ` : null
+    }
+
+    <iron-pages role="main" selected="${page}" attr-for-selected="name" selected-attribute="visible" fallback-selection="404">
+      <!-- home view -->
+      <shop-home name="home"></shop-home>
+      <!-- list view of items in a category -->
+      <shop-list name="list"></shop-list>
+      <!-- detail view of one item -->
+      <shop-detail name="detail"></shop-detail>
+      <!-- cart view -->
+      <shop-cart name="cart"></shop-cart>
+      <!-- checkout view -->
+      <shop-checkout name="checkout"></shop-checkout>
+
+      <shop-404-warning name="404"></shop-404-warning>
+    </iron-pages>
+
+    <footer>
+      <a href="https://www.polymer-project.org/1.0/toolbox/">Made by Polymer</a>
+      <div class="demo-label">Demo Only</div>
+    </footer>
+
+    <!-- a11y announcer -->
+    <div class="announcer" aria-live="assertive">${_a11yLabel}</div>
+
+    ${ modalOpened ? html`<shop-cart-modal></shop-cart-modal>` : null }
+    ${ loadComplete ? html`<shop-snackbar class$="${_snackbarOpened ? 'opened' : ''}">${offline ? 'You are offline' : 'You are online'}</shop-snackbar>` : null }
+    `;
+  }
+
+  static get is() { return 'shop-app'; }
+
+  static get properties() { return {
+    page: String,
+
+    offline: Boolean,
+
+    meta: Object,
+    
+    modalOpened: Object,
+
+    categories: Object,
+
+    categoryName: String,
+
+    drawerOpened: Boolean,
+
+    _a11yLabel: String,
+
+    _smallScreen: Boolean,
+
+    _snackbarOpened: Boolean,
+
+    loadComplete: Boolean
+  }}
+
+  _propertiesChanged(props, changed, oldProps) {
+    if (changed) {
+      if ('categoryName' in changed) {
+        this._categoryNameChanged(props.categoryName, oldProps.categoryName);
+      }
+      if ('page' in changed) {
+        this._pageChanged(props.page, oldProps.page);
+      }
+      if ('offline' in changed) {
+        this._offlineChanged(props.offline, oldProps.offline);
+      }
+      if ('meta' in changed) {
+        this._metaChanged(props.meta, oldProps.meta);
+      }
+    }
+    super._propertiesChanged(props, changed, oldProps);
+  }
+
+  constructor() {
+    super();
+
+    installRouter(() => this._updateLocation());
+    installNetwork(store);
+    store.dispatch(fetchCategories());
+  }
+
+  update() {
+    const state = store.getState();
+
+    this._category = currentCategorySelector(state);
+    this._item = currentItemSelector(state);
+    let page = pageSelector(state);
+    switch (page) {
+      case 'list':
+        if (!this._category) page = '404';
+        break;
+      case 'detail':
+        if (!this._item) page = '404';
```

Comment:
I think just checking `!this._item` is not enough since the app might be still in the process of fetching category items. Maybe also check for `category.loading`, something like this: `!this._category.loading && !this._item`.

Also need to check for if `category` is not null for detail page, for example, if the url is something like this: `http://localhost:8080/detail/foo`